rescript/get_data.py

```diff
@@ -12,13 +12,132 @@
+                'eukaryotes': {False: '10.15156/BIO/2938081', True: '10.15156/BIO/2938082'}},
+        # Old version 9.0 is not listed here
+        '8.3': {'fungi':      {False: '10.15156/BIO/1264708', True: '10.15156/BIO/1264763'},
+                'eukaryotes': {False: '10.15156/BIO/1264819', True: '10.15156/BIO/1264861'}},
+        '8.2': {'fungi':      {False: '10.15156/BIO/786385', True: '10.15156/BIO/786387'},
+                'eukaryotes': {False: '10.15156/BIO/786386', True: '10.15156/BIO/786388'}},
+        '8.0': {'fungi':      {False: '', True: '10.15156/BIO/786349'},
+                'eukaryotes': {False: '', True: ''}},  # All other 8.0 are in zip files
+    }
+    # There's got to be a better way! See https://stackoverflow.com/questions/25833613/safe-method-to-get-value-of-nested-dictionary
+    try:
+        # Check if we have the DOI requested
+        target_doi = unite_dois[version][taxon_group][singletons]
+    except KeyError as ke:
+        print('Unknown DOI for this value: ' + str(ke))
+        raise
+    return _unite_dois_to_urls(target_doi).pop()
+
+_unite_get_url(version='9.0', taxon_group='fungi', singletons=False)
+
+# with tempfile.TemporaryDirectory() as tmp_dir:
+tmp_dir = tempfile.mkdtemp()
+
+def _unite_download_targz(url, download_path):
+    print('Downloading ' + url)
+        
+    response = requests.get(url, stream=True)
+    if response.status_code != 200:
+        raise ValueError("Failed to download the file from " + url)
+    
+    tar_file_path = os.path.join(download_path, 'unitefile.tar.gz')
+    with open(tar_file_path, 'wb') as f:
+        f.write(response.content)
+    
+    # Extract only the 'developer' subdirectory
+    with tarfile.open(tar_file_path, 'r:gz') as tar:
+        # Ensure that 'developer' exists in the tar file
+        members = [member for member in tar.getmembers() if member.name.startswith('developer')]
+        if not members:
+            raise ValueError("No 'developer' subdirectory found in the .tar.gz file.")
+        
+        for member in members:
+            member.name = os.path.basename(member.name) # Strip the 'developer' prefix
+            tar.extract(member, path=download_path)
+    
+    return download_path
+
+# Test it by downloading this file
+# _unite_download_targz('https://files.plutof.ut.ee/public/orig/59/12/591225E8985EFC44B595C79AF5F467421B4D9A95093A0811B13CB4CC13A6DA46.tgz', tmp_dir)
+
+# import as artifacts
+# results[name] = qiime2.Artifact.import_data(dtype, destination)
+
+def get_unite_data(version, taxon_group, singletons=False):
+    url = _unite_get_url(version, taxon_group, singletons)
+    results = {'sequences': [], 'taxonomy': []}
+    
+    # with tempfile.TemporaryDirectory() as tmp_dir:
+    tmp_dir = tempfile.mkdtemp()
+    
+    print('Temporary directory:', tmp_dir)
+    _unite_download_targz(url, download_path=tmp_dir)
+    
+    for root, dirs, files in os.walk(tmp_dir):
```

Comment:
We will not know the exact file names until we download and open the .tar.gz file
(see previous discussion about downloading blind, then importing)
so here, I get file names and import them based on extension.

Is there a better way to do this?

Reply:
I agree, walking is best in this case

Reply:
but do we really not know the file names? is there not any consistent naming scheme?

Reply:
Prefix and suffix are known, though I'm only using suffix here

The middle includes 
- version, taxa-group, and singletons, which we could infer from our DOI dictionary
- date, which is totally unique (this was the other candidate for primary key, except you never know it in advance)

Reply:
You should be able to parse the strings based on this available information. This is exactly what we had to do with SILVA. In fact, we have to adjust the parsers for different versions of SILVA as the naming schema changed for different versions. See [here](https://github.com/bokulich-lab/RESCRIPt/blob/946728467532f066c8bb750c3d8d1b8af02773b9/rescript/get_data.py#L49).